File: module_view/src/main/java/com/lq/view/MainActivity.java
package com.lq.view;

import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        Person[] pers = {null,new Person(1),new Person(2),new Person(3)};
        System.out.println("修改之前=====>");
        for (Person per : pers) {
            System.out.println(per);
        }
        for (int i = 0; i < pers.length; i++) {
            Person per = pers[i];
            if (per!=null) {
                if (per.getAge() ==1) {
                    pers[i] = null;
                }
            }
        }
        System.out.println("修改之后=====>");
        for (Person per : pers) {
            System.out.println(per);
        }
    }
}

File: module_coroutine/src/main/java/com/lq/module_coroutine/MainActivity.kt
package com.lq.module_coroutine

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import kotlinx.coroutines.*

private const val TAG = "MainActivity"

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        Log.i(TAG, "执行前:onCreate: 主线程id:${mainLooper.thread.id}")
        testCoroutine()
        Log.i(TAG, "执行后:onCreate: 主线程id:${mainLooper.thread.id}")
    }


    private fun testCoroutine() {
//        GlobalScope.launch {
//            repeat(8) {
//                Log.i(TAG, "testCoroutine:执行线程id$it:${Thread.currentThread().id}")
//                delay(1000)
//            }
//        }

//        runBlocking {
//            repeat(20) {
//                Log.i(TAG, "testCoroutine:执行线程id$it:${Thread.currentThread().id}")
//                delay(1000)
//            }
//        }

        runBlocking {
//            val job = GlobalScope.launch {
//                println("Throwing exception from launch")
//                throw IndexOutOfBoundsException()
//            }
//
//            job.join()
//            println("Joined failed job")

            val deferred = GlobalScope.async {
                println("Throwing exception from async")
                throw ArithmeticException() // 没有打印任何东西，依赖用户去调用等待
            }
            try {
                deferred.await()
                println("Unreached")
            } catch (e: ArithmeticException) {
                println("Caught ArithmeticException")
            }
        }

    }
}